File: business/src/main/java/work/beltran/discogsbrowser/business/profile/ProfileInteractorImpl.java
package work.beltran.discogsbrowser.business.profile;

import rx.Observable;
import rx.functions.Func1;
import work.beltran.discogsbrowser.api.DiscogsService;
import work.beltran.discogsbrowser.api.model.UserIdentity;
import work.beltran.discogsbrowser.api.model.UserProfile;
import work.beltran.discogsbrowser.business.ProfileInteractor;
import work.beltran.discogsbrowser.business.RxJavaSchedulers;

/**
 * Created by Miquel Beltran on 8/27/16
 * More on http://beltran.work
 */
public class ProfileInteractorImpl implements ProfileInteractor {
    private DiscogsService service;
    private RxJavaSchedulers schedulers;

    public ProfileInteractorImpl(DiscogsService service, RxJavaSchedulers schedulers) {
        this.service = service;
        this.schedulers = schedulers;
    }

    @Override
    public Observable<UserIdentity> getIdentity() {
        return service.getUserIdentity()
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.mainThread());
    }

    @Override
    public Observable<UserProfile> getProfile() {
        return service.getUserIdentity()
                .concatMap(new Func1<UserIdentity, Observable<UserProfile>>() {
                    @Override
                    public Observable<UserProfile> call(UserIdentity userIdentity) {
                        return service.getUserProfile(userIdentity.getUsername())
                                .subscribeOn(schedulers.io())
                                .observeOn(schedulers.mainThread());
                    }
                });
    }
}

File: data-repository/src/main/java/work/beltran/discogsbrowser/business/profile/ProfileInteractorImpl.kt
package work.beltran.discogsbrowser.business.profile

import io.reactivex.Single
import work.beltran.discogsbrowser.api.DiscogsService
import work.beltran.discogsbrowser.api.model.UserIdentity
import work.beltran.discogsbrowser.api.model.UserProfile
import work.beltran.discogsbrowser.business.ProfileInteractor
import work.beltran.discogsbrowser.business.RxJavaSchedulers
import work.beltran.discogsbrowser.business.SettingsRepository

/**
 * Created by Miquel Beltran on 8/27/16
 * More on http://beltran.work
 */
class ProfileInteractorImpl(private val service: DiscogsService,
                            private val settings: SettingsRepository,
                            private val schedulers: RxJavaSchedulers)
    : ProfileInteractor {

    override val identity: Single<UserIdentity>
        get() = service.userIdentity
                .doOnSuccess { settings.setUsername(it.username) }
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.mainThread())

    override val profile: Single<UserProfile>
        get() = service.getUserProfile(settings.getUsername())
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.mainThread())
}
